### PR TITLE
Implement unification of const abstract impls

### PR DESCRIPTION
**Pull request by JulianKnodt**
If two impls cover the entire set of possibilities:

i.e.
```rs
impl TraitName for Struct<true> {}
impl TraitName for Struct<false> {}
```
Then it would be expected that:
```rs
... <const N: usize> ... where Struct<{N > 1}>: TraitName {
```
Should compile.
This allows for such by generating a pseudo-impl for conditions which are exhaustive.

Since these impls will never be instantiated as is, it is fine to have a fake impl.

For now, this checks for a single bool condition, and whether both bool implementations are present.

Not sure who to `r?`but cc @BoxyUwU
Addresses https://github.com/rust-lang/project-const-generics/issues/26

### CODE REVIEW

File: compiler/rustc_trait_selection/src/traits/select/mod.rs
Comment (oli-obk) [inline compiler/rustc_trait_selection/src/traits/select/mod.rs:979]: I think you should not select any candidates in assemble_candidates_from_unified_impls, and instead of bailing out successfully here, actually go into confirmation. Confirmation can then generate obligations for all the variants of the enum/bool

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: compiler/rustc_trait_selection/src/traits/select/mod.rs
Comment (JulianKnodt) [reply compiler/rustc_trait_selection/src/traits/select/mod.rs:979]: So what you're saying is instead of 1 single fake candidates, generate a new obligation that looks for all candidates of the variants?

Comment (oli-obk) [reply compiler/rustc_trait_selection/src/traits/select/mod.rs:979]: I think you'll still want one candidate, so that we first look for a single generic impl with the existing candidates, and if that fails, we process your new candidate in confirmation by spewing an obligation per variant

**Comment by JulianKnodt** [comment]
@oli-obk I've implemented a very sketchy version of what I think you're asking for. If this is the general idea, I'll clean it up and make it a bit more general, but please let me know if this is generally what you're thinking of.

File: compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs
Comment (oli-obk) [inline compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs:216]: I was hoping we could insert it unconditionally if the candidates list is empty after https://github.com/rust-lang/rust/blob/dd12cd6dc631b5e964d541d370ca863c2242376c/compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs#L329-L333

Did you try that?

File: compiler/rustc_trait_selection/src/traits/select/confirmation.rs
Comment (oli-obk) [inline compiler/rustc_trait_selection/src/traits/select/confirmation.rs:1374]: I've been giving this scheme some thought. Instead of generating all possible variants and proving them, maybe we could let the trait solver handle such "questions". Basically you are asking the trait solver whether

* `Type<X>: Trait` can be satisfied (where `X` is a constant)
* which is basically `for<const X: bool> Type<X>: Trait`
* which gets turned into something like `Type<PlaceholderConst>: Trait`
* which we could then treat differently from `Type<ParamConst>` or `Type<true>` or `Type<false>`

When we try to prove that, we could add some special handling for that case that essentially makes a query to list all possible impls for `Type<_>` even if normally it would avoid trying to match inference vars against possible impls. Once we got the list of all impls that match (for bools there should be either one (a generic one) or two (one `true` and one `false`)). I'm not sure this works out, but we should probably explore it before going with a feature that just lists all values. Especially because the listing would get annoying for enums with values

Comment (JulianKnodt) [reply compiler/rustc_trait_selection/src/traits/select/confirmation.rs:1374]: I think this makes sense, as extending this to enums is also a pain. There are also cases which are inductive such as `0usize`, and `N.saturating_sub(1)`. Where would that kind of impl go?

As a counterpoint, this PR was meant to be a small part of this larger feature, but I think most use cases would be for bools and not enums with many variants.

Comment (oli-obk) [reply compiler/rustc_trait_selection/src/traits/select/confirmation.rs:1374]: > I think most use cases would be for bools and not enums with many variants.

I'm not worried about enums with many variants. I was mainly thinking about an impl for `None` and one for `Some(T)`.

I get that this is supposed to be a minimal impl, but I don't know if we should land a minimal impl if we are sure it's going to get removed entirely and replaced by a rewrite (fwiw: I don't think we are sure yet, but I also don't know if anyone'll have the capacity to work on my explorative alternative proposed above).

Comment (JulianKnodt) [reply compiler/rustc_trait_selection/src/traits/select/confirmation.rs:1374]: I can definitely make something which can have more cases added, but I am not entirely sure how to add something like that without explicit checks for certain traits. I've mostly been able to follow the trait selection documentation, since it mostly exists for candidate assembly, but beyond that point there isn't much.

Would it require building or refactoring a large part of the existing obligation system?

Comment (oli-obk) [reply compiler/rustc_trait_selection/src/traits/select/confirmation.rs:1374]: i don't believe so. but i also don't know how to implement it exactly. this would be more exploratory work.

we could do a sync session on eploring the existing implementation and coming up with more concrete ideas.

**Comment by oli-obk** [comment]
I'm on vacation. I will review this in 2-3 weeks

**Comment by JulianKnodt** [comment]
@oli-obk no worries, I've not arrived on a concrete design so am just experimenting

**Comment by albertlarsan68** [comment]
Tests have moved from `src/test` to `tests`, please [rebase](https://rustc-dev-guide.rust-lang.org/git.html#rebasing) on top of new `master`.
@rustbot label -A-bootstrap

**Comment by oli-obk** [comment]
IMO still blocked on trying out https://github.com/rust-lang/rust/pull/104803#discussion_r1033364468

If you want the feature now, you can emulate this PR with specialization, by having a generic dummy default impl and writing specific impls for `true` and `false`

**Comment by JulianKnodt** [comment]
it's been a minute since I've looked at this PR, I'll check it out again but I had mostly forgotten about it due to external stuff

**Comment by onestacked** [comment]
I've done a rebase [here](https://github.com/ink-feather-org/rust/tree/unify_impls).
For some reason including Either caused an error, so i chose to return a `Box<dyn Iterator>` for the return type in `fn exhaustive_types` in confirmation.rs.

`inductive.rs` does not pass, but that was also the case in the current version

**Comment by JulianKnodt** [comment]
@onestacked if you want to take over this PR, that's good with me, but is up to you

**Comment by onestacked** [comment]
I don't know the compiler internals well enough to make actual changes, I mostly rebased to learn a bit of those.

**Comment by JulianKnodt** [comment]
@oli-obk this is the closest thing to what I think you want. For an exhaustive impl, it will add obligations for each implementation for a specific const, rather than exhaustively search for all of them. I didn't add ADTs yet, because I'm not entirely sure how to construct them from valtrees, but it's easy to fill in later.

**Comment by oli-obk** [comment]
> For an exhaustive impl, it will add obligations for each implementation for a specific const, rather than exhaustively search for all of them

This is still the design you basically had originally. You are still iterating over the possible values of `bool` and `i8`.

**Comment by JulianKnodt** [comment]
For that, what I could do is push an additional candidate in `assemble_candidates_from_impls`, where I explicitly replace the predicate's `Const` that is either `Unevaluated | Param` with a `PlaceholderConst`, and then handle it similarly to how I'm handling consts currently in `confirm_exhaustive_candidate`. Ultimately the functionality will be the same (it will still have to exhaustively check types), but it will just be in the same function as the existing user specified impls. Is that more what you're thinking?

**Comment by oli-obk** [comment]
My issue with this approach is that it creates a second exhaustiveness checker that

a) works very differently from the one we have (pattern exhaustiveness collects all patterns and fills out a matrix with the patterns that are there).
b) doesn't clearly allow us to migrate to a scheme that will reuse (or at least share code with) pattern exhaustiveness in the future
c) is "good enough", so if it lands, there is little incentive to remove all of it and create a better version in the future
d) can be emulated with specialization, so I don't think it's useful to have a new very unstable feature just for convenience

I'd be ok without reusing pattern exhaustiveness checking in an initial MVP that we can land, but the approach should be designed in a way that works as if we were using it.

I am very aware how painful that design would be to implement in the old solver, and I don't know how to implement it in the new solver without doing a deep dive myself.

Since it has been lost in the discussion and wasn't very well explained to begin with, I'll try to explain the high level steps that I think should be done, without exactly knowing how the steps can be done currently (they likely need some extra work, or may only work in the new solver)

if you have an predicate of the form `SomeStruct<C>: Trait` to prove, replace the const generic params with inference variables: `SomeStruct<?42c>: Trait`. Then ask the trait solver to collect all candidates that match that predicate. Now you got a list of impls or where bounds or other candidates. This means you are in a situation similar to pattern exhaustiveness checking. Only instead of patterns like `Foo(BAR)` you have types like `SomeStruct<BAR>`, or even `SomeStruct<Some(BAZ)>`. I don't know yet how to generalize exhaustiveness checking to allow you to plug in these types, but I'm sure we could figure that out. But we don't have to initially, we can just do a cheap version that works for booleans and field-less enums similarly to how you did it. Instead of creating obligations for all variants and proving them, you create a list of all the variants that the constant's type has, and then you remove all variants that were found. If there are no variants left, success, we are exhaustive, otherwise you can now report a single error similar to exhaustiveness checking informing the user about which variants are still missing.

**Comment by lcnr** [comment]
> I'd be ok without reusing pattern exhaustiveness checking in an initial MVP that we can land, but the approach should be designed in a way that works as if we were using it.

saw this PR because of https://github.com/rust-lang/rust/pull/113393. Even when using pattern exhaustiveness checking and while being unstable, I would prefer to not currently experiment on this.

Having a kind of working, but not not-appropriate for stabilization, impl here will end up causing issues or conflict in the future. Having something that kinda works makes it harder to rework the surrounding code because we don't want to unnecessarily break peoples projects, even if unstable. It also causes bad vibes™ in the community if we do break it. This has happened when adding `ConstEvaluatable` checks for all constants while working towards the `min_const_generics` stabilization. From what I can tell we also have this issue with `specialization` and `const_trait_impls`.

There are a lot of subtle details here: e.g. should `impl Trait<true> for u32` and `impl<T> Trait<false> for T` should allow us to assume `for<const B: bool> u32: Trait<B>` and how do we check that? This also interacts with how and whether we will consider the following impl to constrain `const N: usize` and therefore be valid: `impl<const N: usize> Trait for Struct<{ Some(N) }>`. I do not want to stabilize anything here until at least stabilizing `generic_const_exprs`, as the design of that will interact with this.

Any such involved changes to the type system should also require a t-types FCP before being landed, even if unstably.

**Comment by tranquillity-codes** [comment]
Ping from triage, @JulianKnodt any updates on resolving the conflicts? Thanks

**Comment by JulianKnodt** [comment]
@tranquillity-codes
It's been a long time since I added this change, I also added a change to another repo which was created to mimic the trait solver, but that has gotten no reviews. Will this PR be reviewed if it's simply rebased?

**Comment by the8472** [comment]
Looks like it has been marked as waiting-for-author since https://github.com/rust-lang/rust/pull/104803#event-8189265971

**Comment by compiler-errors** [comment]
I'm gonna go ahead and close this.

I think T-types is not really free at the moment to take on any major new const generic features, since we're busy working on revamping the base support for const generics (Boxy is leading all that wonderful work!). This PR has sat in an unresolved state for quite some time (sorry about that from the types side!), so I don't want to give the impression that this PR is still open for review sake.

**Pull request closed without merging**
